### ISSUE TITLE
findfont scans all fonts every time

### ISSUE THREAD

**Issue by ederag** [JuliaGraphics/FreeTypeAbstraction.jl#67]
`findfont` works well, but takes several seconds - which may be considered fast, considering - here (openSUSE Leap-15.3),
but led to the significant system-dependent overhead found in https://github.com/cesaraustralia/DynamicGrids.jl/issues/194#issuecomment-995176775.

[`find_font`](https://github.com/JuliaGraphics/FreeTypeAbstraction.jl/blob/6f898ca225a6407a3ca351a89d86d41ef7d6879b/src/findfonts.jl#L125-L171) is always scanning all fonts, opening them
to get their name and properties and find the best one.

The issue is that on my system there were over 10_000 fonts installed, most quick to load (median about 7.5 µs),
but some taking much longer (see the diagnosis histogram), so the average is about 270 µs.

<details>
<summary>diagnosis</summary>

```julia
using FreeTypeAbstraction: fontpaths, try_load, match_font
font_folders = copy(fontpaths())

t1_ = Float64[]
path_ = String[]
@time for folder in font_folders
           for font in readdir(folder)
               fpath = joinpath(folder, font)
               
               t1 = @elapsed face = try_load(fpath)
               push!(t1_, t1)
               push!(path_, fpath)
               face === nothing && continue
               finalize(face)
           end
       end

  2.855721 seconds (159.18 k allocations: 23.692 MiB)

julia> sum(t1_)
2.832251035

julia> using UnicodePlots
julia> histogram(t1_)
                  ┌                                        ┐ 
   [0.0  , 0.001) ┤█████████████████████████████████  10063  
   [0.001, 0.002) ┤▌ 151                                     
   [0.002, 0.003) ┤▍ 83                                      
   [0.003, 0.004) ┤▎ 32                                      
   [0.004, 0.005) ┤▏ 4                                       
   [0.005, 0.006) ┤▏ 14                                      
   [0.006, 0.007) ┤▏ 7                                       
   [0.007, 0.008) ┤▏ 15                                      
   [0.008, 0.009) ┤▏ 9                                       
   [0.009, 0.01 ) ┤▏ 1                                       
   [0.01 , 0.011) ┤▏ 3                                       
   [0.011, 0.012) ┤  0                                       
   [0.012, 0.013) ┤▏ 7                                       
   [0.013, 0.014) ┤▏ 1                                       
   [0.014, 0.015) ┤▏ 1                                       
                  └                                        ┘ 
                                   Frequency                

julia> using Statistics
julia> mean(t1_)
0.00027256770618804736
julia> median(t1_)
7.4702e-5
```
</details>

There are several solutions on the other side (e.g. removing slow fonts, using `FTFont(font_path)` to load the specific font directly).

But it might be nice to have a kind of cache so that `findfont(font_name)` would be even faster on subsequent calls ?

**Comment by rafaqz** [comment]
Thanks for looking into this. DynamicGrids could allow passing in the FTFont directly, but a cache seems like a cleaner solution, and maybe better if we dont all implement that.

Can we just move the cache in Makie here? How easy would it be to just copy the code over? @SimonDanisch @jkrumbiegel

**Comment by ederag** [comment]
Agreed that centralizing would be better.
I opened a [discussion on discourse](https://discourse.julialang.org/t/best-way-to-locate-and-load-a-font/73996) because it might involve [FontConfig.jl](https://github.com/JuliaGraphics/Fontconfig.jl) as well (it's very fast).

**Comment by SimonDanisch** [comment]
Sounds like a good idea to make `findfont` as fast as possible ^^

**Comment by rafaqz** [comment]
Yes depending on FontConfig.jl here does look like the best solution. Im happy to review a PR for this.

**Comment by ederag** [comment]
*If* there is an agreement about making this package dependent on FontConfig.jl, then
I might be able to create a PR, hopefully this week-end.

[`fontconfig` syntax](https://www.freedesktop.org/software/fontconfig/fontconfig-user.html) differs from the current `findfont` one.
So it would be safer to *add* the new function [`load_font`](https://discourse.julialang.org/t/best-way-to-locate-and-load-a-font/73996) described on discourse,
and only after a while, deprecate the current `findfont` that has proven reliable ?

**Comment by SimonDanisch** [comment]
>Yes depending on FontConfig.jl here does look like the best solution.

So I tried pretty hard to use FontConfig instead of rolling our own font search, but I couldn't get it to work reliable on all platforms.
To be honest, I don't really remember the problems anymore, I just remember that I gave up and thought, that a simple findfont would be easier and more reliable.

**Comment by ederag** [comment]
> So I tried pretty hard to use FontConfig instead of rolling our own font search,
> but I couldn't get it to work reliable on all platforms.

https://github.com/JuliaGraphics/Fontconfig.jl/issues/21 and https://github.com/JuliaGraphics/Fontconfig.jl/issues/8
have been solved, but that remains a valid concern, 
as there are two opened issues, both about installation (https://github.com/JuliaGraphics/Fontconfig.jl/issues/12 and https://github.com/JuliaGraphics/Fontconfig.jl/issues/30).
They looked specific to certain configurations, but still good to have in mind.

**Comment by SimonDanisch** [comment]
Seems like there is at least BinaryBuilder now...But:

https://github.com/JuliaGraphics/Fontconfig.jl/pull/31#issuecomment-585247957

And I guess someone will need to maintain Fontconfig.jl and update the CIs etc...

**Comment by timholy** [comment]
This seems relevant for https://github.com/JuliaLang/julia/pull/47184#issuecomment-1364028015. I inserted some debugging code:
```diff
$ git diff
diff --git a/src/findfonts.jl b/src/findfonts.jl
index 0b668a5..fabdc2e 100644
--- a/src/findfonts.jl
+++ b/src/findfonts.jl
@@ -136,9 +136,12 @@ function findfont(

     best_score_so_far = (0, 0, false, typemin(Int))
     best_font = nothing
+    @show font_folders
+    nfonts = 0

     for folder in font_folders
         for font in readdir(folder)
+            nfonts += 1
             fpath = joinpath(folder, font)
             face = try_load(fpath)
             face === nothing && continue
@@ -168,6 +171,7 @@ function findfont(
             end
         end
     end
+    @show nfonts best_font

     return best_font
 end
```

and got this output:

```julia
julia> @time using CairoMakie
 11.008582 seconds (18.42 M allocations: 1.153 GiB, 5.60% gc time, 0.54% compilation time)

julia> @time @eval scatter(0..1, rand(10), markersize=rand(10) .* 20)
font_folders = ["/home/tim/.julia/packages/Makie/Ggejq/assets/fonts", "/usr/share/fonts", "/usr/share/fonts/X11", "/usr/share/fonts/X11/Type1", "/usr/share/fonts/X11/encodings", "/usr/share/fonts/X11/encodings/large", "/usr/share/fonts/X11/misc", "/usr/share/fonts/X11/util", "/usr/share/fonts/cMap", "/usr/share/fonts/cmap", "/usr/share/fonts/cmap/adobe-cns1", "/usr/share/fonts/cmap/adobe-gb1", "/usr/share/fonts/cmap/adobe-japan1", "/usr/share/fonts/cmap/adobe-japan2", "/usr/share/fonts/cmap/adobe-korea1", "/usr/share/fonts/opentype", "/usr/share/fonts/opentype/urw-base35", "/usr/share/fonts/truetype", "/usr/share/fonts/truetype/ancient-scripts", "/usr/share/fonts/truetype/dejavu", "/usr/share/fonts/truetype/droid", "/usr/share/fonts/truetype/lato", "/usr/share/fonts/truetype/noto", "/usr/share/fonts/truetype/unifont", "/usr/share/fonts/type1", "/usr/share/fonts/type1/gsfonts", "/usr/share/fonts/type1/texlive-fonts-recommended", "/usr/share/fonts/type1/urw-base35", "/home/tim/.local/share/fonts", "/usr/local/share/fonts"]
nfonts = 1730
best_font = FTFont (family = TeX Gyre Heros Makie, style = Regular)
font_folders = ["/home/tim/.julia/packages/Makie/Ggejq/assets/fonts", "/usr/share/fonts", "/usr/share/fonts/X11", "/usr/share/fonts/X11/Type1", "/usr/share/fonts/X11/encodings", "/usr/share/fonts/X11/encodings/large", "/usr/share/fonts/X11/misc", "/usr/share/fonts/X11/util", "/usr/share/fonts/cMap", "/usr/share/fonts/cmap", "/usr/share/fonts/cmap/adobe-cns1", "/usr/share/fonts/cmap/adobe-gb1", "/usr/share/fonts/cmap/adobe-japan1", "/usr/share/fonts/cmap/adobe-japan2", "/usr/share/fonts/cmap/adobe-korea1", "/usr/share/fonts/opentype", "/usr/share/fonts/opentype/urw-base35", "/usr/share/fonts/truetype", "/usr/share/fonts/truetype/ancient-scripts", "/usr/share/fonts/truetype/dejavu", "/usr/share/fonts/truetype/droid", "/usr/share/fonts/truetype/lato", "/usr/share/fonts/truetype/noto", "/usr/share/fonts/truetype/unifont", "/usr/share/fonts/type1", "/usr/share/fonts/type1/gsfonts", "/usr/share/fonts/type1/texlive-fonts-recommended", "/usr/share/fonts/type1/urw-base35", "/home/tim/.local/share/fonts", "/usr/local/share/fonts"]
nfonts = 1730
best_font = FTFont (family = TeX Gyre Heros Makie, style = Bold)
 16.750444 seconds (185.45 k allocations: 11.756 MiB, 1.04% compilation time)
FigureAxisPlot()
```

I think you could add a `fonts.jl` to Makie that basically does this:

```julia
const best_regular = findfonts(...)
const best_bold = findfonts(...)
```

and then the choice would be precompiled. (You wouldn't call it at runtime at all.)

**Comment by t-bltg** [comment]
Also, in addition to finding default fonts at compile time as proposed by @timholy, one could cache fonts instead of looking up the font on every [`to_font` invocation](https://github.com/MakieOrg/Makie.jl/blob/a9e4e8f838fc2b61976cd0b5ee4a2ce2defda65c/src/conversions.jl#L939).

This is the [caching mechanism](https://github.com/JuliaPlots/UnicodePlots.jl/blob/master/src/freetype.jl#L86-L101) I wrote for `UnicodePlots`.

**Comment by jkrumbiegel** [comment]
I guess a simple optimization could be just saving the list of font names for all found files in a text file. The font search as it is relies only on family and style name, as I have always found that to be the most reliable way to pick specific font variants. As opposed to try and make the engine match a font whose name I already know by picking weight values etc. I think most users really just want to select specific fonts and do not need a complicated matching engine. So we don't need to open each file just to read family and style name over and over again. The only thing to work out would be when to invalidate the cache.

**Comment by t-bltg** [comment]
I really don't see how we could fill the font_cache at compile time, since a `FTFont` holds a pointer. However, we can cache the font paths as regular string, and avoid to scan directories at runtime.

**Comment by timholy** [comment]
Right, it has to be something durable. Cache the choice, not the result. Reading a single font file will be much faster than reading all of them.
